compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/Analysis/Explorer.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Util functions
+//===----------------------------------------------------------------------===//
+
+static LogicalResult
+foldUnitDimsOnGlobal(IRRewriter &rewriter, IREE::Util::GlobalOpInterface global,
+                     SmallVector<IREE::Util::GlobalLoadOpInterface> loadOps,
+                     SmallVector<IREE::Util::GlobalStoreOpInterface> storeOps,
+                     SymbolTable moduleSymbols) {
+  // Create a new transformed GlobalOp.
+  SmallVector<int64_t> newShape;
+  auto globalType = cast<RankedTensorType>(global.getGlobalType());
+  for (auto size : globalType.getShape()) {
+    if (size != 1) {
+      newShape.push_back(size);
+    }
+  }
+  // Just bail in the special case for tensor of all unit dims.
+  if (newShape.empty()) {
+    return success();
+  }
+  auto newGlobalType =
+      RankedTensorType::get(newShape, globalType.getElementType());
+  std::optional<TypedAttr> newInitialValue = std::nullopt;
+  if (auto uninitializedAttr =
+          llvm::dyn_cast_or_null<IREE::Util::UninitializedAttr>(
+              global.getGlobalInitialValue())) {
+    newInitialValue = IREE::Util::UninitializedAttr::get(rewriter.getContext(),
+                                                         newGlobalType);
+  }
+  // TODO: Handle non-uninitialized cases.
+  else {
+    return success();
+  }
```

Comment:
Let's do it in the other way, which could save level of indents.

```cpp
auto uninitializedAttr = dyn_cast<...>(...);

// TODO...
if (!uninitializedAttr) return success();

newInitialValue = ...
```

Also, is it possible that `getGlobalInitialValue()` is NULL? I think we don't need `or_null` here. And please drop `llvm::`.